### ISSUE TITLE
srs偶发崩溃，由清理无用SrsSource机制造成的

### ISSUE THREAD

**Issue by haofz** [ossrs/srs#713]
加入清理无用SrsSource机制后，
[https://github.com/ossrs/srs/commit/c7b97aa1c3de380451e2d229fc5bef702e2d264d](url)
[https://github.com/ossrs/srs/commit/590e9517398c4d442bcf125421353b4b338217d1](url)
测试中发现srs程序崩溃问题，日志如下：
2016-12-13 14:16:42.790|trace|6104|243876|0|RTMP client ip=192.168.10.20
2016-12-13 14:16:42.792|trace|6104|243876|0|complex handshake success
2016-12-13 14:16:42.792|trace|6104|243876|0|connect app, tcUrl=rtmp://172.20.1.2:1935/live, pageUrl=, swfUrl=, schema=rtmp, vhost=__defaultVhost__, port=1935, app=live, param=, args=null
2016-12-13 14:16:42.792|trace|6104|243876|0|out chunk size to 60000
2016-12-13 14:16:42.872|trace|6104|243876|0|input chunk size to 60000
2016-12-13 14:16:42.873|trace|6104|243876|0|client identified, type=fmle-publish, stream_name=livestream, duration=-1.00
2016-12-13 14:16:42.873|trace|6104|243876|0|update req of soruce for auth ok
2016-12-13 14:16:42.873|trace|6104|243876|0|source url=__defaultVhost__/live/livestream, ip=192.168.10.20, cache=1, is_edge=0, source_id=-1[-1]
2016-12-13 14:16:42.888|trace|6104|243218|0|cleanup die source, total=43
2016-12-13 14:16:42.953|trace|6104|243876|0|>>>>>>>>>>>>> FMLE start to publish stream livestream, url=__defaultVhost__/live/livestream.

**分析原因：**
在一次推流的connect和正式推流之间的时间内有阻塞，该source被clear掉了，程序崩溃。

**过程：**
1、一次推流成功，断开；
2、29秒后二次推流，SrsSource被fetch()出来，fetch()方法中die_at未被置成-1，（此处与create不用）；
3、connect与publishing之间有阻塞，如expect_message()，如果此时正好到30秒，该source会被1s一次的循环清理机制干掉；
4、再使用该SrsSource时就崩溃了；

**解决：**
 SrsSource::fetch()中，加入source->die_at = -1;    个人测试可解决。

**Comment by winlinvip** [comment]
The fetch mechanism of Source is not atomic, this part needs to be changed.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
The problem with it is that after fetching, it may not necessarily create a consumer or publisher. After fetching, if it is not cleaned up, it will cause the source to not be cleaned up (if there are no subsequent publishers or consumers). If the die at status is not changed after fetching, it will result in the source being cleaned up and subsequent use of the source will crash. Therefore, setting die_at=-1 in the fetch is also incorrect.

In other words, the cleanup of the source needs to consider the scenario where it is not used (neither published nor consumed) after fetch_or_create.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
A simpler way is to remove the function of source cleaning.
Support for this feature will be added in the future.

Dup to #1509
Solution: https://github.com/ossrs/srs/issues/1579#issuecomment-736232139

`TRANS_BY_GPT3`

**Comment by darkterrorooo** [comment]
Recently, I have also been researching this memory release issue. In the latest version of srs-2.0.243, I enabled the source cleaning mechanism and conducted the following tests: 

1. I pushed 40 streams concurrently, and the memory reached 60MB. After enabling the source cleaning mechanism and ending the stream, I found that the source was indeed cleaned up. However, the memory still remained at around 60MB.

2. I pushed 100 streams, and the memory reached around 160MB. After ending the stream, I found that the source was also cleaned up. However, the memory will still remain at around 60MB.

So I have a question: Does SRS have a memory retention mechanism? To retain a certain amount of memory to prevent repetitive allocation? I have been searching for a while but couldn't find where it is. I would greatly appreciate it if someone could enlighten me.

Finally, thanks to Winlinvip for their open-source contribution!

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
2.0 has already disabled the cleaning of Source, which will be implemented in 3.0. Is it the modification of the code that caused the crash when enabling source cleaning? Currently, this feature requires more modifications to be supported, so it was disabled in 2.0.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
Dup to #1509

**Comment by winlinvip** [comment]
Fixed by [SmartPtr: Support shared ptr for live source. v6.0.129](https://github.com/ossrs/srs/commit/e7069788e9b628850df35f4efa6b08488bdcccfd)
